std/experimental/color.d

```diff
@@ -0,0 +1,704 @@
+                this.a = a;
+            static if(components.canFind('l'))
+                this.l = toGrayscale!colorSpace(r, g, b);
+        }
+
+        // L/A constructor
+        this(ComponentType l, ComponentType a = 0)
+        {
+            static if(components.canFind('l'))
+                this.l = l;
+            static if(components.canFind('r'))
+                this.r = l;
+            static if(components.canFind('g'))
+                this.g = l;
+            static if(components.canFind('b'))
+                this.b = l;
+            static if(components.canFind('a'))
+                this.a = a;
+        }
+
+        // read hex strings in the standard forms: (#/$/0x)rgb/argb/rrggbb/aarrggbb
+        void fromHex(S)(const(S)[] hex) if(isSomeChar!S)
+        {
+            static ubyte val(S c)
+            {
+                if(c >= 0 && c <= 9)
+                    return c - '0';
+                else if(c >= 'a' && c <= 'f')
+                    return c - 'a' + 10;
+                else if(c >= 'A' && c <= 'F')
+                    return c - 'A' + 10;
+                else
+                    assert(false, "Invalid hex string");
+            }
+
+            if(hex[0] == '#' || hex[0] == '$')
+                hex = hex[1..$];
+            else if(hex[0] == '0' && hex[1] == 'x')
+                hex = hex[2..$];
+
+            if(hex.length == 3)
+            {
+                ubyte r = val(hex[0]);
+                ubyte g = val(hex[1]);
+                ubyte b = val(hex[2]);
+                this = Color!("rgb", ubyte, colorSpace)(r | (r << 4), g | (g << 4), b | (b << 4));
+            }
+            if(hex.length == 4)
+            {
+                ubyte a = val(hex[0]);
+                ubyte r = val(hex[1]);
+                ubyte g = val(hex[2]);
+                ubyte b = val(hex[3]);
+                this = Color!("rgba", ubyte, colorSpace)(r | (r << 4), g | (g << 4), b | (b << 4), a | (a << 4));
+            }
+            if(hex.length == 6)
+            {
+                ubyte r = (val(hex[0]) << 4) | val(hex[1]);
+                ubyte g = (val(hex[2]) << 4) | val(hex[3]);
+                ubyte b = (val(hex[4]) << 4) | val(hex[5]);
+                this = Color!("rgb", ubyte, colorSpace)(r, g, b);
+            }
+            if(hex.length == 8)
+            {
+                ubyte a = (val(hex[0]) << 4) | val(hex[1]);
+                ubyte r = (val(hex[2]) << 4) | val(hex[3]);
+                ubyte g = (val(hex[4]) << 4) | val(hex[5]);
+                ubyte b = (val(hex[6]) << 4) | val(hex[7]);
+                this = Color!("rgba", ubyte, colorSpace)(r, g, b, a);
+            }
+            else
+                assert(false, "Invalid hex string!");
+        }
+    }
+
+    C opCast(C)() if(isColor!C)
+    {
+        return this.to!C;
+    }
+
+    void opBinary(string op)()
+    {
+        static if(op == '-')
+        {
+            this.tupleof = -this.tupleof;
```

Comment:
Wait, this isn't right. Is this binary or unary operator overloading? The function is called `opBinary`, where's the second operand? And why is the `static if` inside the function, and not its template constraint?